# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/nsTextEditorState.cpp
# Commit: 9ee7acab98a4
# Full Hash: 9ee7acab98a470b2860125ba3deb439a64dc83e7
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2019-06-10 21:47:35
# Regressor Bug: 1529884
# File Overlap Count: 1
# Description:
#   Bug 1529884 - part 6: Through subject principal at Document::ExecCommand() to constructor of EditorBase::AutoEditActionDataSetter r=smaug
#   
#   `Document::ExecCommand()` knows subject principal.  This patch makes it tell
#   `EditorCommand::DoCommand()` and `EditorCommand::DoCommandParam()`.  Then,
#   makes they tell each editor public methods which may cause dispatching
# ==============================================================================

diff -r 258c4da663ca -r 9ee7acab98a4 dom/html/nsTextEditorState.cpp
--- a/dom/html/nsTextEditorState.cpp	Mon Jun 10 10:26:04 2019 +0000
+++ b/dom/html/nsTextEditorState.cpp	Mon Jun 10 10:27:07 2019 +0000
@@ -2317,8 +2317,12 @@
             // In this case, we need to dispatch "input" event because
             // web apps may need to know the user's operation.
             RefPtr<nsRange> range;  // See bug 1506439
+            // In this case, we need to dispatch "beforeinput" events since
+            // we're emulating the user's input.  Passing nullptr as
+            // nsIPrincipal means that that may be user's input.  So, let's
+            // do it.
             DebugOnly<nsresult> rv =
-                textEditor->ReplaceTextAsAction(newValue, range);
+                textEditor->ReplaceTextAsAction(newValue, range, nullptr);
             NS_WARNING_ASSERTION(NS_SUCCEEDED(rv),
                                  "Failed to set the new value");
           } else if (aFlags & eSetValue_ForXUL) {
@@ -2350,13 +2354,19 @@
                 StringTail(newValue, newlength - currentLength);
 
             if (insertValue.IsEmpty()) {
+              // In this case, we makes the editor stop dispatching "input"
+              // event so that passing nullptr as nsIPrincipal is safe for
+              // now.
               DebugOnly<nsresult> rv = textEditor->DeleteSelectionAsAction(
-                  nsIEditor::eNone, nsIEditor::eStrip);
+                  nsIEditor::eNone, nsIEditor::eStrip, nullptr);
               NS_WARNING_ASSERTION(NS_SUCCEEDED(rv),
                                    "Failed to remove the text");
             } else {
+              // In this case, we makes the editor stop dispatching "input"
+              // event so that passing nullptr as nsIPrincipal is safe for
+              // now.
               DebugOnly<nsresult> rv =
-                  textEditor->InsertTextAsAction(insertValue);
+                  textEditor->InsertTextAsAction(insertValue, nullptr);
               NS_WARNING_ASSERTION(NS_SUCCEEDED(rv),
                                    "Failed to insert the new value");
             }
@@ -2377,7 +2387,9 @@
               selection->RemoveAllRangesTemporarily();
             }
 
-            textEditor->SetText(newValue);
+            // In this case, we makes the editor stop dispatching "input" event
+            // so that passing nullptr as nsIPrincipal is safe for now.
+            textEditor->SetTextAsAction(newValue, nullptr);
 
             // Call the listener's HandleValueChanged() callback manually, since
             // we don't use the transaction manager in this path and it could be